### PR TITLE
Update of   ZTF and Rubin light curves detection files

### PR DESCRIPTION
**Pull request by weizmannk**
`.  These files are for some, an update and a modification to allow the reading, the simulations of the  light curves  as well as their detections with the ZTF and Rubin Observatory.  In the 'doc' folder tutorial ``gwemopt_light_curves_detection.md`` accompanies this pull request to allow 'splitting' the CBCs of the Farah distribution into BNS, NSBH. As well as the simulation of their detection with ``gwemopt``. 

This tutorial is a start, in the short term we will add the **Hubble constant** simulation procedure used in the paper ``**Updated observing scenarios and projections for H0 during the IGWN's O4 and O5**'' and in the long term the **EOS** from the GW posterior simulation in **Bilby** to the EM posterior in **NMMA**, then to their combination.

### CODE REVIEW

File: nmma/em/detect_lightcurves.py [hunk 1]
```diff
@@ -248,6 +269,8 @@ def main():
     probs_det = probs[idx]
     probs_miss = probs[idy]
 
+    print( dataframe_from_detected["mass_1"], " :",dataframe_from_detected["mass_2"] )
```
Comment (mcoughlin) [inline nmma/em/detect_lightcurves.py:272]: Remove print statements.

File: nmma/em/detect_lightcurves.py [hunk 2]
```diff
@@ -148,19 +150,33 @@ def main():
...
         outdir = os.path.join(args.outdir, str(index))
+        #outdir = os.path.join('./obs_paper/detection_lc/outdir_BNS', str(index))
         if not os.path.isdir(outdir):
             os.makedirs(outdir)
-
+ 
         skymap_file = os.path.join(args.skymap_dir, "%d.fits" % indices[index])
         lc_file = os.path.join(args.lightcurve_dir, "%d.dat" %  indices[index])
+        
+        #skymap_file = os.path.join('~/OBSERVING_SCENARIOS/runs/O4/farah/allsky', "%d.fits" % indices[index])
```
Comment (mcoughlin) [inline nmma/em/detect_lightcurves.py:165]: Should this be an if else statement?

**Review by mcoughlin — APPROVED** [review]
Please remove print and comment statements. Otherwise looks good.

File: nmma/em/detect_lightcurves.py (hunk 2 above)
Comment (weizmannk) [reply nmma/em/detect_lightcurves.py:165]: Do you mean that  if args.ztf_sampling or args.ztf_ToO or args.rubin_ToO ?

Here when we add ztf argument or Rubin, the process break. 
 I remember you suggested to add a print to inform the users.
 
 But we could use try , except ?

Comment (weizmannk) [inline nmma/em/detect_lightcurves.py]:
```suggestion
```

File: nmma/em/detect_lightcurves.py [hunk 3]
```diff
@@ -148,19 +150,33 @@ def main():
 
     #indices = np.loadtxt(args.indices_file)
     indices = simulation_id
-    
+
     commands = []
     lcs = {}
     for index, row in dataframe_from_inj.iterrows():
         outdir = os.path.join(args.outdir, str(index))
+        #outdir = os.path.join('./obs_paper/detection_lc/outdir_BNS', str(index))
```
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py]:
```suggestion
```

File: nmma/em/detect_lightcurves.py [hunk 4]
```diff
@@ -195,19 +211,24 @@ def main():
 
     absmag, effs, probs = [], [], []
     fid = open(args.detections_file, "w")
+    #fid = open('lc_skymap_detection', "w")
     for index, row in dataframe_from_inj.iterrows():
         outdir = os.path.join(args.outdir, str(index))
+        #outdir = os.path.join('./obs_paper/detection_lc/outdir_BNS', str(index))
```
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py:216]:
```suggestion
```

File: nmma/em/detect_lightcurves.py [hunk 5]
```diff
@@ -248,6 +269,8 @@ def main():
     probs_det = probs[idx]
     probs_miss = probs[idy]
 
+    print( dataframe_from_detected["mass_1"], " :",dataframe_from_detected["mass_2"] )
+    print( dataframe_from_missed)
```
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py]:
```suggestion
```

File: nmma/em/detect_lightcurves.py (hunk 2 above)
Comment (mcoughlin) [reply nmma/em/detect_lightcurves.py:165]: Yeah try except and raise an error.

File: nmma/em/detect_lightcurves.py [hunk 6]
```diff
@@ -148,19 +150,33 @@ def main():
...
+        
+        #lcs[index] = np.loadtxt(lc_file)[0:number_shot,]
+        
         lcs[index] = np.loadtxt(lc_file)
+        
+        print(lcs)
+  
 
         efffile = os.path.join(outdir, f"efficiency_true_{indices[index]}.txt")
+        print(efffile)
```
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py:179]:
```suggestion
```

File: nmma/em/detect_lightcurves.py [hunk 7]
```diff
@@ -195,19 +211,24 @@ def main():
 
     absmag, effs, probs = [], [], []
     fid = open(args.detections_file, "w")
+    #fid = open('lc_skymap_detection', "w")
     for index, row in dataframe_from_inj.iterrows():
         outdir = os.path.join(args.outdir, str(index))
+        #outdir = os.path.join('./obs_paper/detection_lc/outdir_BNS', str(index))
+        
         efffile = os.path.join(outdir, f"efficiency_true_{indices[index]}.txt")
+        #print(efffile)
```
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py:216]:
```suggestion
```

File: nmma/em/detect_lightcurves.py [hunk 8]
```diff
@@ -195,19 +211,24 @@ def main():
...
         # for ZTF r-band give the best proxy 
         if args.telescope == 'ZTF':
             absmag.append(np.min(lcs[index][:, 3]))
-        # For Rubin i-band or z-band are best but here we choose r-band,
-        # that give the same results  
+        # For Rubin i-band or z-band are best 
         else:
             absmag.append(np.min(lcs[index][:, 4]))
             
+        print(absmag)
```
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py:229]:
```suggestion
```

File: nmma/em/detect_lightcurves.py (hunk 1 above)
Comment (weizmannk) [inline nmma/em/detect_lightcurves.py:268]:
```suggestion
```

File: nmma/em/detect_lightcurves.py (hunk 2 above)
Comment (weizmannk) [reply nmma/em/detect_lightcurves.py]: You do you mean I could raise simply  the error  ( except as err )  or 
return the error  as an information contain in the print (expcet  print("..........") ?

Comment (mcoughlin) [reply nmma/em/detect_lightcurves.py]: Sure.

Comment (weizmannk) [reply nmma/em/detect_lightcurves.py]: Ok, but, 
the erro when we use ztf-sampling , is 
IndexError: index 1 is out of bounds for axis 0 with size 1

Si I am wondering if here try except ,  could not be  
hide (mask) other similar errors ?

Comment (mcoughlin) [reply nmma/em/detect_lightcurves.py]: Maybe first check for IndexError then other exceptions?

File: nmma/em/create_lightcurves.py [hunk 9]
```diff
@@ -189,212 +200,221 @@ def main():
...
+        # from observing scenarios 
+        simulation_id = injection_df['simulation_id']
+        
+        mag_ds = {}
+        for index, row in injection_df.iterrows():
+            
+            injection_outfile = os.path.join(args.outdir, "%d.dat" % simulation_id[index])
+            if os.path.isfile(injection_outfile):
+                mag_ds[index] = np.loadtxt(injection_outfile)
+                continue
```
Comment (weizmannk) [inline nmma/em/create_lightcurves.py:300]:
```suggestion
            try:
                mag_ds[index] = np.loadtxt(injection_outfile)
                continue
                
            except ValueError :
                print('\n===================================================================') 
                print('The previous run generated light curves with unreadable content.\n')
                print('Please remove all output files in .dat format then retry.')
                print('===================================================================\n')
                exit()

```

File: nmma/em/create_lightcurves.py [hunk 10]
```diff
@@ -189,212 +200,221 @@ def main():
...
+            for ii, tt in enumerate(sample_times):
+                fid.write("%.5f " % sample_times[ii])
+                for filt in data.keys():
+                    if args.filters:
+                        if filt not in args.filters.split(","):
+                            continue
+                    fid.write("%.3f " % data[filt][ii, 1])
+                fid.write("\n")
+            fid.close()
+
```
Comment (weizmannk) [inline nmma/em/create_lightcurves.py:333]:
```suggestion

        try:
            fid = open(injection_outfile, "w")
            # fid.write('# t[days] u g r i z y J H K\n')
            # fid.write(str(" ".join(('# t[days]'," ".join(args.filters.split(',')),"\n"))))
            fid.write("# t[days] ")
            fid.write(str(" ".join(args.filters.split(","))))
            fid.write("\n")

            for ii, tt in enumerate(sample_times):
                fid.write("%.5f " % sample_times[ii])
                for filt in data.keys():
                    if args.filters:
                        if filt not in args.filters.split(","):
                            continue
                    fid.write("%.3f " % data[filt][ii, 1])
                fid.write("\n")
            fid.close()
            
        except IndexError:
            print('\n===================================================================') 
            print('Sorry we could not use ZTF or Rubin flags to generate those statistics\n')
            print('Please remove all flags rely on with ZTF or Rubin then retry again')
            print('===================================================================\n')
            exit()
```

**Comment by weizmannk** [comment]
It's Ok now

**Pull request merged**